Ensure the password reset template is correctly converted to binary

babolivier opened the pull request:
Python 2 seems to choke when trying to convert a file's content to bytes if that content includes special characters. This looks like it's because `.read()` on a file descriptor returns a `str` instead of a `unicode` on Python 2, which makes `content.encode("utf8")` fail because it doesn't expect a special char.

Python 2:

```
Python 2.7.16 (default, Mar 11 2019, 18:59:25) 
[GCC 8.2.1 20181127] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> f = open("/home/brendan/Documents/matrix/synapse/synapse/res/templates/account_validity_renewal_success.html")
>>> s = f.read()
>>> type(s)
<type 'str'>
>>> s.encode("utf8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 536: ordinal not in range(128)
```

Python 3:

```
Python 3.7.4 (default, Jul 16 2019, 07:12:58) 
[GCC 9.1.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> f = open("/home/brendan/Documents/matrix/synapse/synapse/res/templates/account_validity_renewal_success.html")
>>> s = f.read()
>>> type(s)
<class 'str'>
>>> s.encode("utf8")
b'<!DOCTYPE html>\n<html lang="fr-FR">\n<head>\n  <meta charset="utf-8"/>\n  <meta http-equiv="Cache-control" content="no-cache">\n\n  <style type="text/css">\n    body {\n      font-family: Helvetica, Arial, Sans-Serif;\n    }\n    #head {\n      font-size:32px;\n      text-align:center;\n    }\n    #msg {\n      margin-top: 20px;\n      font-size: 25px;\n      text-align: center;\n    }\n  </style>\n\n</head>\n<body>\n<div id="head">\n  <img id="logo" src="/images/tchap192.png" width="192" height="192" alt="Tchap" />\n</div>\n\n<div id="msg">Votre compte a \xc3\xa9t\xc3\xa9 renouvel\xc3\xa9 avec succ\xc3\xa8s.</div>\n</body>\n</html>\n'
```

This would cause Synapse to respond with a 500 error to any client trying to query one of the affected endpoints if the HTML to serve contains special characters (e.g. french translation of the existing files).

six's `ensure_binary` function ensures that what we try to send to the client is of the right type (which is `str` on Python 2 and `bytes` on Python 3). FWIW, it's already used, working and tested on this fork since https://github.com/matrix-org/synapse/pull/5932.

babolivier committed:
Ensure the password reset template is correctly converted to binary

Regardless of the Python version

babolivier committed:
Changelog

richvdh commented:
>  This looks like it's because .read() on a file descriptor returns a str instead of a unicode on Python 2,

correct, python2's `open`/`read` return the raw bytes, so it is incorrect to try to do a `.encode` on them. The solution to that is to either read the file as binary on both python 2 and 3 (with `open(..., "rb")`) or to read it as text on both python 2 and 3 (by giving it an explicit encoding with `open(..., encoding="utf-8")`). See https://python-future.org/compatible_idioms.html#file-io-with-open.

richvdh commented on synapse/rest/client/v2_alpha/account.py:
```diff
@@ -295,7 +296,7 @@ def on_GET(self, request, medium):
             request.setResponseCode(e.code)
 
-        request.write(html.encode('utf-8'))
+        request.write(ensure_binary(html))
```
sorry, but this still doesn't look right. the bug here is that `html` is having an identity crisis over whether it contains unicode or bytes. Presumably that's happening because the things you're assigning to html are having similar identity crises.

The solution here isn't to go round sticking plasters pver that problem every time we spit out some html: rather it is to make sure that things are consistently unicode or bytes.

In other words: this needs doing earlier on. Presumably the problem is where `email_password_reset_success_html_content` is read from disk. go and fix it there instead.

babolivier replied:
Ah, right, I understand now. Roger that, I'll look into it.

babolivier committed:
Revert "Ensure the password reset template is correctly converted to binary"

This reverts commit 665dd9f7f8db2f8d38f454d4d5b96efcf163a5db.

babolivier committed:
Read all files as UTF-8

babolivier committed:
Revert "Merge pull request #5932 from matrix-org/babolivier/account_validity_template_encode"

This reverts commit 84e695f506faf54982b9e19dceb9c02acffad95f, reversing
changes made to 99eec6d2d5cc76e645c3fd7ca6cda85b2bab6feb.

babolivier committed:
Lint

babolivier commented:
I've updated the PR so that it doesn't rely on `ensure_binary` and instead uses `io.open` with the `utf-8` encoding.

I've also reverted https://github.com/matrix-org/synapse/pull/5932 because this change made it redundant and because we don't need `six>=1.12` anymore. Given https://github.com/matrix-org/synapse/pull/5932 hasn't been deployed in production yet, I believe that's something we can do.

richvdh reviewed (dismissed):
tests are failing :-p

babolivier committed:
Rename io.open import to limite side-effects

richvdh reviewed (approved):
lgtm!

I wouldn't be *entirely* surprised to discover that some other bit of code is relying on `read_file` returning a `bytes`, but said code would be broken on python 3, so let's hope not...
